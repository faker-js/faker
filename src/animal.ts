import type { Faker } from '.';

export class Animal {
  constructor(private readonly faker: Faker) {
    // Bind `this` so namespaced is working correctly
    for (const name of Object.getOwnPropertyNames(Animal.prototype)) {
      if (name === 'constructor' || typeof this[name] !== 'function') {
        continue;
      }
      this[name] = this[name].bind(this);
    }
  }

  /**
   * Returns a random dog breed.
   *
   */
  dog(): string {
    return this.faker.random.arrayElement(this.faker.definitions.animal.dog);
  }

  /**
   * Returns a random cat breed.
   *
   */
  cat(): string {
    return this.faker.random.arrayElement(this.faker.definitions.animal.cat);
  }

  /**
   * Returns a random snake species.
   *
   */
  snake(): string {
    return this.faker.random.arrayElement(this.faker.definitions.animal.snake);
  }

  /**
   * Returns a random bear species.
   *
   */
  bear(): string {
    return this.faker.random.arrayElement(this.faker.definitions.animal.bear);
  }

  /**
   * Returns a random lion species.
   *
   */
  lion(): string {
    return this.faker.random.arrayElement(this.faker.definitions.animal.lion);
  }

  /**
   * Returns a random cetacean species.
   *
   * @method faker.animal.cetacean
   */
  cetacean(): string {
    return this.faker.random.arrayElement(
      this.faker.definitions.animal.cetacean
    );
  }

  /**
   * Returns a random horse breed.
   *
   */
  horse(): string {
    return this.faker.random.arrayElement(this.faker.definitions.animal.horse);
  }

  /**
   * Returns a random bird species.
   *
   */
  bird(): string {
    return this.faker.random.arrayElement(this.faker.definitions.animal.bird);
  }

  /**
   * Returns a random cow species.
   *
   */
  cow(): string {
    return this.faker.random.arrayElement(this.faker.definitions.animal.cow);
  }

  /**
   * Returns a random fish species.
   *
   */
  fish(): string {
    return this.faker.random.arrayElement(this.faker.definitions.animal.fish);
  }

  /**
   * Returns a random crocodilian species.
   *
   */
  crocodilia(): string {
    return this.faker.random.arrayElement(
      this.faker.definitions.animal.crocodilia
    );
  }

  /**
   * Returns a random insect species.
   *
   */
  insect(): string {
    return this.faker.random.arrayElement(this.faker.definitions.animal.insect);
  }

  /**
   * Returns a random rabbit species
   *
   */
  rabbit(): string {
    return this.faker.random.arrayElement(this.faker.definitions.animal.rabbit);
  }

  /**
   * Returns a random animal type.
   *
   */
  type(): string {
    return this.faker.random.arrayElement(this.faker.definitions.animal.type);
  }
}
