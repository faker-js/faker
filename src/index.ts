import { Faker } from './faker';
import allLocales from './locales';

export type {
  AnimalDefinitions,
  ColorDefinitions,
  CommerceDefinitions,
  CommerceProductNameDefinitions,
  CompanyDefinitions,
  DatabaseDefinitions,
  DateDefinitions,
  DateEntryDefinition,
  FinanceCurrencyEntryDefinitions,
  FinanceDefinitions,
  HackerDefinitions,
  InternetDefinitions,
  LocaleDefinition,
  /** @deprecated Use LocationDefinitions instead */
  LocationDefinitions as AddressDefinitions,
  LocationDefinitions,
  LoremDefinitions,
  MusicDefinitions,
  /** @deprecated Use PersonDefinitions instead */
  PersonDefinitions as NameDefinitions,
  PersonDefinitions,
  /** @deprecated Use PersonTitleDefinitions instead */
  PersonTitleDefinitions as NameTitleDefinitions,
  PersonTitleDefinitions,
  PhoneNumberDefinitions,
  ScienceDefinitions,
  SystemDefinitions,
  SystemMimeTypeEntryDefinitions,
  VehicleDefinitions,
  WordDefinitions,
} from './definitions';
export { FakerError } from './errors/faker-error';
export type { FakerOptions, UsableLocale, UsedLocales } from './faker';
export type { AnimalModule } from './modules/animal';
export type {
  Casing,
  ColorFormat,
  ColorModule,
  CSSFunction,
  CSSSpace,
  NumberColorFormat,
  StringColorFormat,
} from './modules/color';
export type { CommerceModule } from './modules/commerce';
export type { CompanyModule } from './modules/company';
export type { DatabaseModule } from './modules/database';
export type { DatatypeModule } from './modules/datatype';
export type { DateModule } from './modules/date';
export type { FinanceModule } from './modules/finance';
export type { GitModule } from './modules/git';
export type { HackerModule } from './modules/hacker';
export type { HelpersModule } from './modules/helpers';
export type { ImageModule } from './modules/image';
export type { InternetModule } from './modules/internet';
export type {
  /** @deprecated Use LocationModule instead */
  LocationModule as AddressModule,
  LocationModule,
} from './modules/location';
export type { LoremModule } from './modules/lorem';
export type { MusicModule } from './modules/music';
export { Sex } from './modules/person';
export type {
  /** @deprecated Use PersonModule instead */
  PersonModule as NameModule,
  PersonModule,
  SexType,
} from './modules/person';
export type { PhoneModule } from './modules/phone';
export type { RandomModule } from './modules/random';
export type { ChemicalElement, ScienceModule, Unit } from './modules/science';
export type { StringModule } from './modules/string';
export type { SystemModule } from './modules/system';
export type { VehicleModule } from './modules/vehicle';
export type { WordModule } from './modules/word';
export { Faker };

// since we are requiring the top level of faker, load all locales by default
export const faker: Faker = new Faker({
  locales: allLocales,
});
