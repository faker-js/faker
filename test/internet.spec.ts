import validator from 'validator';
import { afterEach, describe, expect, it } from 'vitest';
import { faker } from '../src';
import { seededTests } from './support/seededRuns';
import { times } from './support/times';

const NON_SEEDED_BASED_RUN = 5;

describe('internet', () => {
  afterEach(() => {
    faker.locale = 'en';
  });

  seededTests(faker, 'internet', (t) => {
    t.itEach(
      'avatar',
      'protocol',
      'httpMethod',
      'domainName',
      'domainSuffix',
      'domainWord',
      'ip',
      'ipv4',
      'ipv6',
      'port',
      'userAgent'
    );

    t.describe('email', (t) => {
      t.it('noArgs')
        .it('with names', 'Jane', 'Doe')
        .it('with provider', undefined, undefined, 'fakerjs.dev')
        .it('with names and provider', 'Jane', 'Doe', 'fakerjs.dev');
    });

    t.describe('exampleEmail', (t) => {
      t.it('noArgs').it('with names', 'Jane', 'Doe');
    });

    t.describe('userName', (t) => {
      t.it('noArgs').it('with names', 'Jane', 'Doe');
    });

    t.describe('password', (t) => {
      t.it('noArgs').it('with length', 10);
    });

    t.describe('httpStatusCode', (t) => {
      t.it('noArgs').it('with options', { types: ['clientError'] });
    });

    t.describe('color', (t) => {
      t.it('noArgs').it('with color base', 100, 100, 100);
    });

    t.describe('mac', (t) => {
      t.it('noArgs').it('with separator', ':');
    });

    t.describe('emoji', (t) => {
      t.it('noArgs').it('with options', { types: ['nature'] });
    });

    t.describe('url', (t) => {
      t.it('noArgs')
        .it('with slash appended', { appendSlash: true })
        .it('without slash appended and with http protocol', {
          appendSlash: false,
          protocol: 'http',
        });
    });
  });

  describe(`random seeded tests for seed ${faker.seed()}`, () => {
    for (let i = 1; i <= NON_SEEDED_BASED_RUN; i++) {
      describe('avatar', () => {
        it('should return a random avatar url', () => {
          const avatar = faker.internet.avatar();

          expect(avatar).toBeTruthy();
          expect(avatar).toBeTypeOf('string');
          expect(avatar).toSatisfy(validator.isURL);
          expect(avatar).toMatch(
            /^https:\/\/cloudflare-ipfs.com\/ipfs\/Qmd3W5DuhgHirLHGVixi6V76LhCkZUz6pnFt5AJBiyvHye\/avatar\/\d+.jpg$/
          );
        });
      });

      describe('email()', () => {
        it('should return an email', () => {
          const email = faker.internet.email();

          expect(email).toBeTruthy();
          expect(email).toBeTypeOf('string');
          expect(email).toSatisfy(validator.isEmail);

          const [, suffix] = email.split('@');
          expect(faker.definitions.internet.free_email).toContain(suffix);
        });

        it('should return an email with given firstName', () => {
          const email = faker.internet.email('Aiden.Harann55');

          expect(email).toBeTruthy();
          expect(email).toBeTypeOf('string');
          expect(email).toSatisfy(validator.isEmail);

          const [prefix, suffix] = email.split('@');

          expect(prefix).includes('Aiden.Harann55');
          expect(prefix).toMatch(
            /^(Aiden\.Harann55((\d{1,2})|([._][A-Za-z]*(\d{1,2})?)))/
          );
          expect(faker.definitions.internet.free_email).toContain(suffix);
        });

        it('should return an email with given firstName and lastName', () => {
          const email = faker.internet.email('Aiden', 'Harann');

          expect(email).toBeTruthy();
          expect(email).toBeTypeOf('string');
          expect(email).toSatisfy(validator.isEmail);

          const [prefix, suffix] = email.split('@');

          expect(prefix).includes('Aiden');
          expect(prefix).toMatch(
            /^Aiden((\d{1,2})|([._]Harann\d{1,2})|([._](Harann)))/
          );
          expect(faker.definitions.internet.free_email).toContain(suffix);
        });

        it('should return an email with japanese characters', () => {
          const email = faker.internet.email('思源_唐3');

          expect(email).toBeTruthy();
          expect(email).toBeTypeOf('string');
          expect(email).toSatisfy(validator.isEmail);

          const [prefix, suffix] = email.split('@');

          expect(prefix).toMatch(/^思源_唐3/);
          expect(faker.definitions.internet.free_email).toContain(suffix);
        });

        it('should return an email with special characters', () => {
          const email = faker.internet.email('Mike', 'Smith', null, {
            allowSpecialCharacters: true,
          });

          expect(email).toBeTruthy();
          expect(email).toBeTypeOf('string');
          expect(email).toSatisfy(validator.isEmail);

          const [prefix, suffix] = email.split('@');

          expect(prefix).toMatch(
            /^Mike((\d{1,2})|([.!#$%&'*+-/=?^_`{|}~]Smith\d{1,2})|([.!#$%&'*+-/=?^_`{|}~]Smith))/
          );
          expect(faker.definitions.internet.free_email).toContain(suffix);
        });
      });

      describe('exampleEmail()', () => {
        it('should return an email with the example suffix', () => {
          const email = faker.internet.exampleEmail();

          expect(email).toBeTruthy();
          expect(email).toBeTypeOf('string');
          expect(email).toSatisfy(validator.isEmail);

          const suffix = email.split('@')[1];

          expect(suffix).toMatch(/^example\.(com|net|org)$/);
          expect(faker.definitions.internet.example_email).toContain(suffix);
        });

        it('should return an email with the example suffix and given firstName', () => {
          const email = faker.internet.exampleEmail('Aiden.Harann55');

          expect(email).toBeTruthy();
          expect(email).toBeTypeOf('string');
          expect(email).toSatisfy(validator.isEmail);

          const [prefix, suffix] = email.split('@');

          expect(suffix).toMatch(/^example\.(com|net|org)$/);
          expect(faker.definitions.internet.example_email).toContain(suffix);
          expect(prefix).toMatch(/^Aiden.Harann55/);
        });

        it('should return an email with the example suffix and given firstName and lastName', () => {
          const email = faker.internet.exampleEmail('Aiden', 'Harann');

          expect(email).toBeTruthy();
          expect(email).toBeTypeOf('string');
          expect(email).toSatisfy(validator.isEmail);

          const [prefix, suffix] = email.split('@');

          expect(suffix).toMatch(/^example\.(com|net|org)$/);
          expect(faker.definitions.internet.example_email).toContain(suffix);
          expect(prefix).toMatch(/^Aiden([._]Harann)?\d*/);
        });

        it('should return an email with the example suffix and japanese characters', () => {
          const email = faker.internet.exampleEmail('思源_唐3');

          expect(email).toBeTruthy();
          expect(email).toBeTypeOf('string');
          expect(email).toSatisfy(validator.isEmail);

          const [prefix, suffix] = email.split('@');

          expect(suffix).toMatch(/^example\.(com|net|org)$/);
          expect(faker.definitions.internet.example_email).toContain(suffix);
          expect(prefix).toMatch(/^思源_唐3/);
        });

        it('should return an email with special characters', () => {
          const email = faker.internet.exampleEmail('Mike', 'Smith', {
            allowSpecialCharacters: true,
          });

          expect(email).toBeTruthy();
          expect(email).toBeTypeOf('string');
          expect(email).toSatisfy(validator.isEmail);

          const [prefix, suffix] = email.split('@');

          expect(suffix).toMatch(/^example\.(com|net|org)$/);
          expect(faker.definitions.internet.example_email).toContain(suffix);
          expect(prefix).toMatch(/^Mike([.!#$%&'*+-/=?^_`{|}~]Smith)?\d*/);
        });
      });

      describe('userName()', () => {
        it('should return a random username', () => {
          const username = faker.internet.userName();

          expect(username).toBeTruthy();
          expect(username).toBeTypeOf('string');
          expect(username).toMatch(/\w/);
        });

        it('should return a random username with given firstName', () => {
          const username = faker.internet.userName('Aiden');

          expect(username).toBeTruthy();
          expect(username).toBeTypeOf('string');
          expect(username).toMatch(/\w/);
          expect(username).includes('Aiden');
        });

        it('should return a random username with given firstName and lastName', () => {
          const username = faker.internet.userName('Aiden', 'Harann');

          expect(username).toBeTruthy();
          expect(username).toBeTypeOf('string');
          expect(username).includes('Aiden');
          expect(username).toMatch(
            /^Aiden((\d{1,2})|([._]Harann\d{1,2})|([._](Harann)))/
          );
        });
      });

      describe('protocol()', () => {
        it('should return a valid protocol', () => {
          const protocol = faker.internet.protocol();
          expect(protocol).toBeTruthy();
          expect(protocol).toBeTypeOf('string');
          expect(protocol).toMatch(/^https?$/);
        });
      });

      describe('httpMethod()', () => {
        const httpMethods = [
          'GET',
          'POST',
          'PUT',
          'DELETE',
          'PATCH',
          'HEAD',
          'OPTIONS',
        ];

        it('should return a valid http method', () => {
          const httpMethod = faker.internet.httpMethod();

          expect(httpMethod).toBeTruthy();
          expect(httpMethod).toBeTypeOf('string');
          expect(httpMethods).toContain(httpMethod);
        });
      });

      describe('httpStatusCode', () => {
        it('should return a random HTTP status code', () => {
          const httpStatusCode = faker.internet.httpStatusCode();

          expect(httpStatusCode).toBeTruthy();
          expect(httpStatusCode).toBeTypeOf('number');
          expect(httpStatusCode).toBeLessThanOrEqual(600);
        });

        it('should return a correct status code for multiple classes', () => {
          const httpStatusCode = faker.internet.httpStatusCode({
            types: ['informational', 'success', 'redirection'],
          });

          expect(httpStatusCode).toBeTruthy();
          expect(httpStatusCode).toBeTypeOf('number');
          expect(httpStatusCode).toBeGreaterThanOrEqual(100);
          expect(httpStatusCode).toBeLessThan(400);
        });

        it('should return a correct status code for a single class', () => {
          const httpStatusCode = faker.internet.httpStatusCode({
            types: ['serverError'],
          });

          expect(httpStatusCode).toBeTruthy();
          expect(httpStatusCode).toBeTypeOf('number');
          expect(httpStatusCode).toBeGreaterThanOrEqual(500);
          expect(httpStatusCode).toBeLessThan(600);
        });
      });

      describe('url()', () => {
        it('should return a valid url', () => {
          const url = faker.internet.url();

          expect(url).toBeTruthy();
          expect(url).toBeTypeOf('string');
          expect(url).toSatisfy(validator.isURL);
        });

        it('should return a valid url with slash appended at the end', () => {
          const url = faker.internet.url({ appendSlash: true });

          expect(url).toBeTruthy();
          expect(url).toBeTypeOf('string');
          expect(url).toSatisfy(validator.isURL);
          expect(url.endsWith('/')).toBeTruthy();
        });

        it('should return a valid url with given protocol', () => {
          const url = faker.internet.url({ protocol: 'http' });

          expect(url).toBeTruthy();
          expect(url).toBeTypeOf('string');
          expect(url).toSatisfy(validator.isURL);
        });
      });

      describe('domainName()', () => {
        it('should return a domainWord plus a random suffix', () => {
          const domainName = faker.internet.domainName();

          expect(domainName).toBeTruthy();
          expect(domainName).toBeTypeOf('string');
          expect(domainName).toSatisfy(validator.isFQDN);

          const [prefix, suffix] = domainName.split('.');

          expect(prefix).toSatisfy(validator.isSlug);
          expect(faker.definitions.internet.domain_suffix).toContain(suffix);
        });
      });

      describe('domainSuffix', () => {
        it('should return a random domainSuffix', () => {
          const domainSuffix = faker.internet.domainSuffix();

          expect(domainSuffix).toBeTruthy();
          expect(domainSuffix).toBeTypeOf('string');
          expect(faker.definitions.internet.domain_suffix).toContain(
            domainSuffix
          );
        });
      });

      describe('domainWord()', () => {
        it('should return a lower-case adjective + noun', () => {
          const domainWord = faker.internet.domainWord();

          expect(domainWord).toBeTruthy();
          expect(domainWord).toBeTypeOf('string');
          expect(domainWord).toSatisfy(validator.isSlug);
          expect(domainWord).toSatisfy((value: string) =>
            validator.isFQDN(value, { require_tld: false })
          );
        });
      });

      describe('ip()', () => {
        it('should return a random IPv4 or IPv6 address', () => {
          const ip = faker.internet.ip();

          expect(ip).toBeTruthy();
          expect(ip).toBeTypeOf('string');
          expect(ip).toSatisfy(validator.isIP);
        });
      });

      describe('ipv4()', () => {
        it('should return a random IPv4 with four parts', () => {
          const ip = faker.internet.ipv4();

          expect(ip).toBeTruthy();
          expect(ip).toBeTypeOf('string');
          expect(ip).toSatisfy((value: string) => validator.isIP(value, 4));

          const parts = ip.split('.');

          expect(parts).toHaveLength(4);

          for (const part of parts) {
            expect(part).toMatch(/^\d+$/);
            expect(+part).toBeGreaterThanOrEqual(0);
            expect(+part).toBeLessThanOrEqual(255);
          }
        });
      });

      describe('ipv6()', () => {
        it('should return a random IPv6 address with eight parts', () => {
          const ipv6 = faker.internet.ipv6();

          expect(ipv6).toBeTruthy();
          expect(ipv6).toBeTypeOf('string');
          expect(ipv6).toSatisfy((value: string) => validator.isIP(value, 6));

          const parts = ipv6.split(':');

          expect(parts).toHaveLength(8);
        });
      });

      describe('port()', () => {
        it('should return a random port number', () => {
          const port = faker.internet.port();

          expect(port).toBeTypeOf('number');
          expect(port).toBeGreaterThanOrEqual(0);
          expect(port).toBeLessThanOrEqual(65535);
          expect(String(port)).toSatisfy(validator.isPort);
        });
      });

      describe('userAgent()', () => {
        it('should return a valid user-agent', () => {
          const ua = faker.internet.userAgent();

          expect(ua).toBeTruthy();
          expect(ua).toBeTypeOf('string');
          expect(ua.length).toBeGreaterThanOrEqual(1);
          expect(ua).toMatch(
            /^(([^\d]+\/[\dA-Za-z\.]+(\s\(.*\)))|([^\d]+\/[\dA-Za-z\.]+(\s\(.*\)*))(\s[^\d]+\/[\dA-Za-z\.]+(\s\(.*\)*))*)$/
          );
        });
      });

      describe('color()', () => {
        it('should return a random hex value', () => {
          const color = faker.internet.color();

          expect(color).toBeTruthy();
          expect(color).toBeTypeOf('string');
          expect(color).toSatisfy(validator.isHexColor);
        });

        it('should return a random hex value with given values', () => {
          const color = faker.internet.color(100, 100, 100);

          expect(color).toBeTruthy();
          expect(color).toBeTypeOf('string');
          expect(color).toSatisfy(validator.isHexColor);
        });
      });

      describe('mac()', () => {
        it('should return a random MAC address with 6 hexadecimal digits', () => {
          const mac = faker.internet.mac();

          expect(mac).toBeTruthy();
          expect(mac).toBeTypeOf('string');
          expect(mac).toHaveLength(17);
          expect(mac).toMatch(/^([a-f0-9]{2}:){5}[a-f0-9]{2}$/);
          expect(mac).toSatisfy(validator.isMACAddress);
        });

        it('should return a random MAC address with 6 hexadecimal digits and given separator', () => {
          const mac = faker.internet.mac('-');

          expect(mac).toBeTruthy();
          expect(mac).toBeTypeOf('string');
          expect(mac).toHaveLength(17);
          expect(mac).toMatch(/^([a-f0-9]{2}-){5}[a-f0-9]{2}$/);
          expect(mac).toSatisfy(validator.isMACAddress);
        });

        it('should return a random MAC address with 6 hexadecimal digits and empty separator', () => {
          const mac = faker.internet.mac('');

          expect(mac).toBeTruthy();
          expect(mac).toBeTypeOf('string');
          expect(mac).toSatisfy(validator.isHexadecimal);
          expect(mac).toHaveLength(12);
          // It's not a valid MAC address
        });

        it.each(['!', '&', '%', '?', '$'])(
          "uses the default (':') if we provide an unacceptable separator ('%s')",
          (separator) => {
            const mac = faker.internet.mac(separator);

            expect(mac).toBeTruthy();
            expect(mac).toBeTypeOf('string');
            expect(mac).toHaveLength(17);
            expect(mac).toMatch(/^([a-f0-9]{2}:){5}[a-f0-9]{2}$/);
            expect(mac).toSatisfy(validator.isMACAddress);
          }
        );
      });

      describe('password', () => {
        it('should return random password', () => {
          const password = faker.internet.password();

          expect(password).toBeTruthy();
          expect(password).toBeTypeOf('string');
          expect(password).toHaveLength(15);
          expect(password).toMatch(/^\w{15}$/);
        });

        it.each(times(32))(
          'should return random password with length %i',
          (length) => {
            const password = faker.internet.password(length);

            expect(password).toBeTruthy();
            expect(password).toBeTypeOf('string');
            expect(password).toHaveLength(length);
            expect(password).toMatch(/^\w+$/);
          }
        );

        it('should return memorable password', () => {
          const password = faker.internet.password(12, true);

          expect(password).toBeTruthy();
          expect(password).toBeTypeOf('string');
          expect(password).toHaveLength(12);
          expect(password).toMatch(/^\w{12}$/);
        });

        it('should return non memorable password', () => {
          const password = faker.internet.password(12, false);

          expect(password).toBeTruthy();
          expect(password).toBeTypeOf('string');
          expect(password).toHaveLength(12);
          expect(password).toMatch(/^\w{12}$/);
          // TODO @Shinigami92 2022-02-11: I would say a non memorable password should satisfy `validator.isStrongPassword`, but it does not currently
          //expect(password).toSatisfy(validator.isStrongPassword);
        });

        it('should return non memorable strong password with length 32', () => {
          const password = faker.internet.password(
            32,
            false,
            /(!|\?|&|\[|\]|%|\$|[a-zA-Z0-9])/
          );

          expect(password).toBeTruthy();
          expect(password).toBeTypeOf('string');
          expect(password).toHaveLength(32);
          // TODO @Shinigami92 2022-02-11: This should definitely be a strong password, but it doesn't :(
          //expect(password).toSatisfy(validator.isStrongPassword);
        });

        it('should return non memorable strong password with length 32 and given prefix', () => {
          const password = faker.internet.password(
            32,
            false,
            /(!|\?|&|\[|\]|%|\$|[a-zA-Z0-9])/,
            'a!G6'
          );

          expect(password).toBeTruthy();
          expect(password).toBeTypeOf('string');
          expect(password).toHaveLength(32);
          expect(password).toMatch(/^a!G6/);
          expect(password).toSatisfy(validator.isStrongPassword);
        });
      });

      describe('emoji', () => {
        it('should return a random emoji', () => {
          const emoji = faker.internet.emoji();

          expect(emoji).toBeTruthy();
          expect(emoji).toBeTypeOf('string');
          expect(emoji.length).toBeGreaterThanOrEqual(1);
        });
      });
    }
  });
});
