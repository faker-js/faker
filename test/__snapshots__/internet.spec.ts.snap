// Vitest Snapshot v1

exports[`internet > 42 > avatar 1`] = `"https://cloudflare-ipfs.com/ipfs/Qmd3W5DuhgHirLHGVixi6V76LhCkZUz6pnFt5AJBiyvHye/avatar/468.jpg"`;

exports[`internet > 42 > color > noArgs 1`] = `"#30667a"`;

exports[`internet > 42 > color > with color base 1`] = `"#6298ac"`;

exports[`internet > 42 > domainName 1`] = `"hasty-shin.org"`;

exports[`internet > 42 > domainSuffix 1`] = `"info"`;

exports[`internet > 42 > domainWord 1`] = `"hasty-shin"`;

exports[`internet > 42 > email > noArgs 1`] = `"Peyton73@yahoo.com"`;

exports[`internet > 42 > email > with names 1`] = `"Jane_Doe18@yahoo.com"`;

exports[`internet > 42 > email > with names and provider 1`] = `"Jane_Doe@fakerjs.dev"`;

exports[`internet > 42 > email > with provider 1`] = `"Garnet.Schinner73@fakerjs.dev"`;

exports[`internet > 42 > emoji > noArgs 1`] = `"🕸️"`;

exports[`internet > 42 > emoji > with options 1`] = `"🦔"`;

exports[`internet > 42 > exampleEmail > noArgs 1`] = `"Peyton73@example.com"`;

exports[`internet > 42 > exampleEmail > with names 1`] = `"Jane_Doe18@example.com"`;

exports[`internet > 42 > httpMethod 1`] = `"POST"`;

exports[`internet > 42 > httpStatusCode > noArgs 1`] = `207`;

exports[`internet > 42 > httpStatusCode > with options 1`] = `410`;

exports[`internet > 42 > ip 1`] = `"cf2b:c992:7210:7d59:2ba0:0fbd:f302:f294"`;

exports[`internet > 42 > ipv4 1`] = `"95.203.243.46"`;

exports[`internet > 42 > ipv6 1`] = `"5cf2:bc99:2721:07d5:92ba:00fb:df30:2f29"`;

exports[`internet > 42 > mac > noArgs 1`] = `"5c:f2:bc:99:27:21"`;

exports[`internet > 42 > mac > with separator 1`] = `"5c:f2:bc:99:27:21"`;

exports[`internet > 42 > password > noArgs 1`] = `"Dl2fkYYKLsZdepz"`;

exports[`internet > 42 > password > with length 1`] = `"Dl2fkYYKLs"`;

exports[`internet > 42 > port 1`] = `24545`;

exports[`internet > 42 > protocol 1`] = `"http"`;

exports[`internet > 42 > url > noArgs 1`] = `"https://staid-veil.biz"`;

exports[`internet > 42 > url > with slash appended 1`] = `"https://hasty-shin.org/"`;

exports[`internet > 42 > url > without slash appended and with http protocol 1`] = `"http://hasty-shin.org"`;

exports[`internet > 42 > userAgent 1`] = `"Mozilla/5.0 (Windows; U; Windows NT 6.2) AppleWebKit/538.0.2 (KHTML, like Gecko) Chrome/29.0.815.0 Safari/538.0.2"`;

exports[`internet > 42 > userName > noArgs 1`] = `"Garnet.Schinner73"`;

exports[`internet > 42 > userName > with names 1`] = `"Jane_Doe"`;

exports[`internet > 1211 > avatar 1`] = `"https://cloudflare-ipfs.com/ipfs/Qmd3W5DuhgHirLHGVixi6V76LhCkZUz6pnFt5AJBiyvHye/avatar/1160.jpg"`;

exports[`internet > 1211 > color > noArgs 1`] = `"#773a72"`;

exports[`internet > 1211 > color > with color base 1`] = `"#a96ca4"`;

exports[`internet > 1211 > domainName 1`] = `"vibrant-infix.org"`;

exports[`internet > 1211 > domainSuffix 1`] = `"org"`;

exports[`internet > 1211 > domainWord 1`] = `"vibrant-infix"`;

exports[`internet > 1211 > email > noArgs 1`] = `"Jadyn.Trantow12@hotmail.com"`;

exports[`internet > 1211 > email > with names 1`] = `"Jane_Doe@hotmail.com"`;

exports[`internet > 1211 > email > with names and provider 1`] = `"Jane.Doe89@fakerjs.dev"`;

exports[`internet > 1211 > email > with provider 1`] = `"Tito_Koelpin22@fakerjs.dev"`;

exports[`internet > 1211 > emoji > noArgs 1`] = `"🇮🇸"`;

exports[`internet > 1211 > emoji > with options 1`] = `"🌲"`;

exports[`internet > 1211 > exampleEmail > noArgs 1`] = `"Jadyn.Trantow12@example.net"`;

exports[`internet > 1211 > exampleEmail > with names 1`] = `"Jane_Doe@example.net"`;

exports[`internet > 1211 > httpMethod 1`] = `"PATCH"`;

exports[`internet > 1211 > httpStatusCode > noArgs 1`] = `505`;

exports[`internet > 1211 > httpStatusCode > with options 1`] = `429`;

exports[`internet > 1211 > ip 1`] = `"117.228.199.57"`;

exports[`internet > 1211 > ipv4 1`] = `"237.117.228.199"`;

exports[`internet > 1211 > ipv6 1`] = `"e7ec:32f0:a2a3:c652:bbd0:caab:de64:dfdf"`;

exports[`internet > 1211 > mac > noArgs 1`] = `"e7:ec:32:f0:a2:a3"`;

exports[`internet > 1211 > mac > with separator 1`] = `"e7:ec:32:f0:a2:a3"`;

exports[`internet > 1211 > password > noArgs 1`] = `"yLuj60b5iHB0bhn"`;

exports[`internet > 1211 > password > with length 1`] = `"yLuj60b5iH"`;

exports[`internet > 1211 > port 1`] = `60851`;

exports[`internet > 1211 > protocol 1`] = `"https"`;

exports[`internet > 1211 > url > noArgs 1`] = `"https://jubilant-temple.net/"`;

exports[`internet > 1211 > url > with slash appended 1`] = `"https://vibrant-infix.org/"`;

exports[`internet > 1211 > url > without slash appended and with http protocol 1`] = `"http://vibrant-infix.org"`;

exports[`internet > 1211 > userAgent 1`] = `"Mozilla/5.0 (Macintosh; U; Intel Mac OS X 10_5_6 rv:3.0; PT) AppleWebKit/538.2.0 (KHTML, like Gecko) Version/4.1.0 Safari/538.2.0"`;

exports[`internet > 1211 > userName > noArgs 1`] = `"Tito_Koelpin22"`;

exports[`internet > 1211 > userName > with names 1`] = `"Jane.Doe89"`;

exports[`internet > 1337 > avatar 1`] = `"https://cloudflare-ipfs.com/ipfs/Qmd3W5DuhgHirLHGVixi6V76LhCkZUz6pnFt5AJBiyvHye/avatar/327.jpg"`;

exports[`internet > 1337 > color > noArgs 1`] = `"#214814"`;

exports[`internet > 1337 > color > with color base 1`] = `"#537a46"`;

exports[`internet > 1337 > domainName 1`] = `"fair-migration.com"`;

exports[`internet > 1337 > domainSuffix 1`] = `"biz"`;

exports[`internet > 1337 > domainWord 1`] = `"fair-migration"`;

exports[`internet > 1337 > email > noArgs 1`] = `"Kellen27@gmail.com"`;

exports[`internet > 1337 > email > with names 1`] = `"Jane.Doe@gmail.com"`;

exports[`internet > 1337 > email > with names and provider 1`] = `"Jane56@fakerjs.dev"`;

exports[`internet > 1337 > email > with provider 1`] = `"Devyn21@fakerjs.dev"`;

exports[`internet > 1337 > emoji > noArgs 1`] = `"💇🏼‍♀️"`;

exports[`internet > 1337 > emoji > with options 1`] = `"🐪"`;

exports[`internet > 1337 > exampleEmail > noArgs 1`] = `"Kellen27@example.org"`;

exports[`internet > 1337 > exampleEmail > with names 1`] = `"Jane.Doe@example.org"`;

exports[`internet > 1337 > httpMethod 1`] = `"POST"`;

exports[`internet > 1337 > httpStatusCode > noArgs 1`] = `205`;

exports[`internet > 1337 > httpStatusCode > with options 1`] = `407`;

exports[`internet > 1337 > ip 1`] = `"8234:8705:3894:5f4b:41c6:1a52:bf27:dccc"`;

exports[`internet > 1337 > ipv4 1`] = `"67.143.40.54"`;

exports[`internet > 1337 > ipv6 1`] = `"4823:4870:5389:45f4:b41c:61a5:2bf2:7dcc"`;

exports[`internet > 1337 > mac > noArgs 1`] = `"48:23:48:70:53:89"`;

exports[`internet > 1337 > mac > with separator 1`] = `"48:23:48:70:53:89"`;

exports[`internet > 1337 > password > noArgs 1`] = `"9V05TL7RY9fmECg"`;

exports[`internet > 1337 > password > with length 1`] = `"9V05TL7RY9"`;

exports[`internet > 1337 > port 1`] = `17172`;

exports[`internet > 1337 > protocol 1`] = `"http"`;

exports[`internet > 1337 > url > noArgs 1`] = `"https://neat-chopstick.biz"`;

exports[`internet > 1337 > url > with slash appended 1`] = `"https://fair-migration.com/"`;

exports[`internet > 1337 > url > without slash appended and with http protocol 1`] = `"http://fair-migration.com"`;

exports[`internet > 1337 > userAgent 1`] = `"Mozilla/5.0 (Windows; U; Windows NT 6.1) AppleWebKit/532.0.0 (KHTML, like Gecko) Chrome/13.0.832.0 Safari/532.0.0"`;

exports[`internet > 1337 > userName > noArgs 1`] = `"Devyn21"`;

exports[`internet > 1337 > userName > with names 1`] = `"Jane56"`;
